hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyLifecycleService.java

```diff
@@ -366,175 +372,230 @@ private void evaluateFSOBucket(OmVolumeArgs volume, OmBucketInfo bucket, String
+              StringBuffer secondLastDirPath = new StringBuffer();
+              for (int i = 0; i < dirList.size() - 1; i++) {
+                secondLastDirPath.append(dirList.get(i).getName());
+                if (i != dirList.size() - 2) {
+                  secondLastDirPath.append(OM_KEY_PREFIX);
+                }
+              }
+              objPath = secondLastDirPath.toString();
+            } else {
+              objID = bucket.getObjectID();
+              objPrefix = OM_KEY_PREFIX + volume.getObjectID() + OM_KEY_PREFIX + bucket.getObjectID() +
+                  OM_KEY_PREFIX + bucket.getObjectID();
+              objPath = "";
+            }
+            try {
+              List<OmDirectoryInfo> subDirList = getSubDirectory(objID, objPrefix, omMetadataManager);
+              for (OmDirectoryInfo subDir : subDirList) {
+                String subDirPath = objPath.isEmpty() ? subDir.getName() : objPath + OM_KEY_PREFIX + subDir.getName();
+                if (!subDir.getName().equals(TRASH_PREFIX) && subDirPath.startsWith(rule.getEffectivePrefix()) &&
+                    (lastDir == null || subDir.getObjectID() != lastDir.getObjectID())) {
+                  evaluateKeyAndDirTable(bucket, volume.getObjectID(), keyTable, subDirPath, subDir,
+                      Arrays.asList(rule), expiredKeyList, expiredDirList);
+                }
+              }
+            } catch (IOException e) {
+              // log failure and continue the process
+              LOG.warn("Failed to get sub directories of {} under {}/{}", objPrefix,
+                  bucket.getVolumeName(), bucket.getBucketName(), e);
+              return;
             }
-            expiredDirList.add(directoryPath.toString(), 0, lastDir.getUpdateID());
           }
+        } else {
+          evaluateKeyAndDirTable(bucket, volume.getObjectID(), keyTable, "", null,
+              Arrays.asList(rule), expiredKeyList, expiredDirList);
         }
+      }
+
+      if (!noPrefixRuleList.isEmpty()) {
+        evaluateKeyAndDirTable(bucket, volume.getObjectID(), keyTable, "", null,
+            noPrefixRuleList, expiredKeyList, expiredDirList);
+      }
+    }
 
-        LOG.info("Prefix {} for {}", prefix, bucketKey);
-        evaluateKeyTable(keyTable, prefix, directoryPath.toString(), rule, expiredKeyList, bucket);
-        evaluateDirTable(directoryInfoTable, prefix, directoryPath.toString(), rule,
-            expiredDirList, bucket);
+    @SuppressWarnings("checkstyle:parameternumber")
+    private void evaluateKeyAndDirTable(OmBucketInfo bucket, long volumeObjId, Table<String, OmKeyInfo> keyTable,
+        String directoryPath, OmDirectoryInfo dir, List<OmLCRule> ruleList,
+        LimitedExpiredObjectList keyList, LimitedExpiredObjectList dirList) {
+      String volumeName = bucket.getVolumeName();
+      String bucketName = bucket.getBucketName();
+      Deque<PendingEvaluateDirectory> stack = new ArrayDeque<>();
```

Comment:
It might be better to check the deepest subdirectory first, but currently it seems to be traversing from the top to the bottom.